includes/class-wc-payments-payment-request-button-handler.php

```diff
@@ -211,6 +211,22 @@ public function custom_button_selector() {
 		return $this->is_custom_button() ? '#wcpay-custom-button' : '';
 	}
 
+	/**
+	 * Gets the product total price.
+	 *
+	 * @param object $product WC_Product_* object.
+	 * @return integer Total price.
+	 */
+	public function get_product_price( $product ) {
+		$product_price = $product->get_price();
+		// Add subscription sign-up fees to product price.
+		if ( 'subscription' === $product->get_type() && class_exists( 'WC_Subscriptions_Product' ) ) {
+			$product_price = $product->get_price() + WC_Subscriptions_Product::get_sign_up_fee( $product );
```

Comment:
I'm not an expert here but a bit concerned by this approach. Apparently if fixes particular issue but it doesn't look very extensible to me. I mean there might be other fees or plugins adding their own fees which ideally would be included into total. So that makes me think there should be some WC Core solution to calculate total (at least something is already used for cart) and I'd imagine we reuse it rather than making our own.

Reply:
That's a good point and I also tried to find another WC core way of getting the product price other than `$product->get_price()`, but the sign up fee is a special thing from WC Subscriptions which is not part of the recurring price. [It's artificially added to the price on the first purchase](https://github.com/woocommerce/woocommerce-subscriptions/blob/82c9d287d35a87bb487559ac76d05d7b71f9a4b0/includes/class-wc-subscriptions-cart.php#L214-L221). I'd love to know if there's a better way of doing this, but I didn't find any generic way of getting the total including sign-up fees by looking at the source.

We offer a [limited number](https://github.com/Automattic/woocommerce-payments/blob/c435726fb9a50be9a3016bdcf1e4b5b21c8be08d/includes/class-wc-payments-payment-request-button-handler.php#L464-L478) of compatible product types, and some of them require special logic for the product page. Like [variable products](https://github.com/Automattic/woocommerce-payments/blob/c435726fb9a50be9a3016bdcf1e4b5b21c8be08d/includes/class-wc-payments-payment-request-button-handler.php#L242-L261).

We do take advantage of the cart totals when the payment request button is displayed on the cart or checkout page, or before the purchase is confirmed.

`get_product_price` is used only for displaying the price on the product page for the first time. - We get the final cart price once the payment is confirmed by adding the item to the cart and displaying the cart totals including tax and shipping once an address is chosen in the payment request dialog.

Reply:
I asked Lambda (p1620838312163000-slack-CJTS771PB) and they say there's no core way of getting the total price for the PRB from the product page without doing this sort of special logic for subscription products.

We are working on a more robust fix for this in Stripe through adopting a different approach  in 1563-gh-woocommerce/woocommerce-gateway-stripe.

We shall sync the fix here later, but for now I think we can consider this PR as a temporary fix.